src/AdsClient/AmsSocketConnection.cs

```diff
@@ -78,33 +80,32 @@ private async Task Listen()
 
     private async Task<byte[]> ListenForHeader()
     {
-        var buffer = new byte[AmsHeaderHelper.AmsTcpHeaderSize];
-        await ReceiveAsync(buffer);
-
-        return buffer;
+        await ReceiveAsync(headerBuffer);
+        return headerBuffer;
     }
 
     private async Task ReceiveAsync(byte[] buffer)
     {
-        using var args = new SocketAsyncEventArgs();
-        args.SetBuffer(buffer, 0, buffer.Length);
-        var awaitable = new SocketAwaitable(args);
+        var sa = socketAwaitable;
+        sa.SetBuffer(buffer, 0, buffer.Length);
 
         do
         {
-            args.SetBuffer(args.Offset + args.BytesTransferred, args.Count - args.BytesTransferred);
-            await socket.ReceiveAsync(awaitable);
+            sa.SetBuffer(sa.Offset + sa.BytesTransferred, sa.Count - sa.BytesTransferred);
+            await socket.ReceiveAwaitable(sa);
 
-            if (args.BytesTransferred == 0)
+            if (sa.BytesTransferred == 0)
             {
                 messageHandler.HandleException(new Exception("Remote host closed the connection."));
                 Close();
             }
-        } while (args.Count != args.BytesTransferred);
+        } while (socketAwaitable.BytesTransferred != buffer.Length);
     }
 
     private async Task ReceiveLoop()
     {
+        await Task.Yield();
```

Comment:
Instead of the `Task.Run` in the ctor of this class.

Reply:
`Task.Run` was borrowed from RabbitMQ.Client, but I like this one as well.